semgrep-core/Analyzing/Bloom_annotation.ml

```diff
@@ -53,23 +91,94 @@ module B = Bloom_filter
  * See also Rules_filter.reserved_id and reserved_str and
  * Rules_filter.extract_specific_strings.
 *)
-let extract_strings _any =
-  []
+
+let rec statement_strings stmt =
+  let res = ref [] in
+  let top_level = ref true in
+  let visitor = V.mk_visitor {
+    V.default_visitor with
+    V.kident = (fun (_k, _) (str, _tok) ->
+      if not (special_ident str) then
+        push str res
+    );
+    V.kexpr = (fun (k, _) x ->
+      (match x with
+       (* less: we could extract strings for the other literals too?
+        * atoms, chars, even int?
+       *)
+       | L (String (str, _tok)) ->
+           if not (special_literal str) then push str res
+       | TypedMetavar _ -> ()
+       | _ -> k x
+      )
+    );
+    V.kconstness = (fun (k, _) x ->
+      (match x with
+       | Lit (String (str, _tok)) ->
+           if not (special_literal str) then push str res
+       | _ -> k x
+      )
+    );
+    V.kid_info = (fun (k, _) x -> k x);
+    V.kstmt = (fun (k, _) x ->
+      (* First statement visited is the current statement *)
+      if !top_level then (top_level := false; k x)
```

Comment:
why do we do something special for the top statement?
why we don't set its bloom filter too?

Reply:
This is because of how I visit so that I can get all the strings in a statement (including its substatements). I use the visitor until I come to a statement S. Then, I stop that visitor, and recursively call the statement_strings function, which starts a ref for a list of strings and creates a new visitor for S. However, S was already visited by the previous visitor, so it should not be revisited. I couldn't quite figure out how to do this with only one visitor call, though I'm afraid this will use more memory.